paasta_tools/config_utils.py

```diff
@@ -0,0 +1,198 @@
+KNOWN_CONFIG_TYPES = ("marathon", "kubernetes")
+HEADER_COMMENT = """
+# This file contains recommended config values for your service generated by
+# automated processes.
+#
+# Your service will use these values if they are not defined in
+# {regular_filename}.
+#
+# If you would like to override a config value defined here, add the config
+# value to {regular_filename} instead of editing this file.
+# ==============================================================================
+{{}}
+"""
+# ^ Needs an empty dict at the end for ruamel to return a non-None value when loading
+# ^ Braces are doubled for escaping in call to .format
+
+
+def write_auto_config_data(
+    service: str, extra_info: str, data: Dict[str, Any], soa_dir: str = DEFAULT_SOA_DIR
+) -> Optional[str]:
+    """
+    Replaces the contents of an automated config file for a service, or creates the file if it does not exist.
+
+    Returns the filename of the modified file, or None if no file was written.
+    """
+    service_dir = f"{soa_dir}/{service}"
+    if not os.path.exists(service_dir):
+        log.warning(
+            f"Service {service} does not exist in configs, skipping auto config update"
+        )
+        return None
+    subdir = f"{service_dir}/{AUTO_SOACONFIG_SUBDIR}"
+    if not os.path.exists(subdir):
+        os.mkdir(subdir)
+    filename = f"{subdir}/{extra_info}.yaml"
+    with open(filename, "w") as f:
+        content = yaml.round_trip_load(
+            HEADER_COMMENT.format(regular_filename=f"{service}/{extra_info}.yaml")
+        )
+        content.update(data)
+        f.write(yaml.round_trip_dump(content))
+    return filename
+
+
+def _commit_files(files: List[str], message: str) -> bool:
+    """
+    Stages the given files and creates a commit with the given message.
+
+    Returns True if a new commit was created, False if the files are unchanged.
+    """
+    subprocess.check_call(["git", "add"] + files)
+    # Skip commit if no changes are staged
+    result_code = subprocess.call(["git", "diff-index", "--quiet", "--cached", "HEAD"])
+    if result_code == 0:
+        return False
+    else:
+        subprocess.check_call(["git", "commit", "--no-verify", "--message", message])
+        return True
+
+
+class PushNotFastForwardError(Exception):
+    pass
+
+
+def _push_to_remote(branch: str) -> None:
+    try:
+        subprocess.check_output(
+            ("git", "push", "origin", branch), stderr=subprocess.STDOUT
+        )
+    except subprocess.CalledProcessError as e:
+        if "Updates were rejected" in str(e.stdout):
+            raise PushNotFastForwardError()
+        else:
+            log.error(f"Push to {branch} failed with:\n {e.stdout}")
+            raise
+
+
+def validate_auto_config_file(filepath: str):
+    basename = os.path.basename(filepath)
+    for file_type in KNOWN_CONFIG_TYPES:
+        if basename.startswith(file_type):
+            return bool(validate_schema(filepath, f"auto/{file_type}"))
```

Comment:
Is this still 'auto'? Not 'autotuning_defaults'? Maybe we should put the name into a constant?

Reply:
It's just the path for the schemas, so it doesn't matter too much - I didn't think we needed to use the full name for internal stuff. But I can definitely use the constant from paasta_tools.utils if you think that would be better.

Reply:
Yes, I'd use a constant just in case we change our mind and need to rename the directory and if I to revisit this code in a half of a year it would be easier to connect logically related parts if the naming is consistent. Thanks!